### PR TITLE
Recipes rework

### LINKED ISSUE #6: Improve items and recipes models
By now, we have all these things written in stone in the `Crafty` contract. 
Can we model a set of items and a set of recipes handled by the `Crafty` owner?
We may need to have a `Recipe` interface as we did for the `Item`s

### LINKED ISSUE #7: Improve items to work as ERC20 tokens
`Item`s are just `Ownable` and carry out their own logic, we should improve them to be ERC20 tokens.

### PR DESCRIPTION
**Pull request by nventuro**
Instead of having two public functions per item (one to make new items, one to retrieve the current amount), the `Crafty` contract now has two public functions total (for the same purpose) that receive an item id as a parameter. The recipe logic is stored in a data structure in the contract's storage.

A couple comments:
* The design would be a bit cleaner if the recipes (or recipe-based items) were their own contract, holding the rules to craft themselves, but this poses a problem: to be able to call the `Item.subtract` method, they'd need to own the `Item` contract, which is owned by `Crafty`. Having an ownership model with multiple owners (or one owner and multiple admins) would solve this, but I'm not sure it's worth it. Once `Item` is an ERC20 token, I could also call `approve` with a huge amount, but this feels hacky.

* I considered having the ids be Solidity enums, but they are somewhat annoying to use right now because [they cannot currently be keys of a mapping](https://solidity.readthedocs.io/en/develop/types.html#mappings), nor [can their values be accessed in JS](https://solidity.readthedocs.io/en/develop/frequently-asked-questions.html#if-i-return-an-enum-i-only-get-integer-values-in-web3-js-how-to-get-the-named-values). The enums have to be casted to an integer type if they are to be used in a mapping, and a manual mapping of item to id would have to be included in `rules.json`, which would unnecessarily complicate things. In the end, I opted for using strings as an id for now.

* Most of `Crafty` could be abstracted away into an owned `CraftableItems` contract, with no baked-in recipes, since they'd be added dynamically after deployment. I'm sticking with static recipes for the time being, since the game is as of now static, and adding the game recipes in a migration or some such mechanism would greatly complicate the testing of the final game contract.

* `CraftableItems` could still be added, and `Crafty` could simply add the recipes in it's constructor, but that wouldn't provide any benefit right now (other than arguably cleaner code), and it increases the gas cost of deploying `Crafty` (to the point that `truffle`'s default gas limit is excedeed).

Suggestions are welcome!

Closes #6 
Closes #7

### CODE REVIEW

**Comment by spalladino** [comment]
I really like how this is looking! A few comments:

> The design would be a bit cleaner if the recipes (or recipe-based items) were their own contract, holding the rules to craft themselves, but this poses a problem: to be able to call the Item.subtract method, they'd need to own the Item contract, which is owned by Crafty. Having an ownership model with multiple owners (or one owner and multiple admins) would solve this, but I'm not sure it's worth it. *Once Item is an ERC20 token, I could also call approve with a huge amount, but this feels hacky.*

This is actually a pretty standard practice. Many decentralized exchanges rely on this hack, where the token owners approve very large amounts to an exchange contract, so it can manage all tokens on their behalf. There is even a proposal for [implementing unlimited allowance in ERC20](https://github.com/ethereum/EIPs/issues/717).

My suggestion would be to make items ERC20s, and use the _huge amount approval_ hack (where a huge amount can be `MAX_UINT_256`).

As for the enums, agree with the decision, they still have a long way to go in Solidity.

Regarding to accepting recipes in the constructor, let's revisit that after the recipes become contracts themselves. In that case, it may end up being cleaner after all.

A last minor comment: there are a few comments in the code on when/how to call a function (such as not calling it with the same id twice). In those cases, it is a good practice to add such restrictions as `require`s, even if it spends a bit more gas.

**Comment by nventuro** [comment]
Nice, I did not know about that proposal, thanks!

> My suggestion would be to make items ERC20s, and use the huge amount approval hack (where a huge amount can be MAX_UINT_256).

In that case, I'll get that done first and then merge here, so that this PR covers the whole refactor.

> A last minor comment: there are a few comments in the code on when/how to call a function (such as not calling it with the same id twice). In those cases, it is a good practice to add such restrictions as `require`s, even if it spends a bit more gas.

I thought about this, but wasn't sure how to check for a zero-valued contract to make sure the slot in the mapping is empty. I didn't find much info in how contracts are actually stored, but it looks like they are pretty much a derived type from address, so checking against a null address might work. Otherwise, I could simply store the actual addresses and then cast to contract when needed (though I'm not sure if this would cost gas or not).

**Comment by spalladino** [comment]
> I thought about this, but wasn't sure how to check for a zero-valued contract to make sure the slot in the mapping is empty. I didn't find much info in how contracts are actually stored, but it looks like they are pretty much a derived type from address, *so checking against a null address might work*. Otherwise, I could simply store the actual addresses and then cast to contract when needed (though I'm not sure if this would cost gas or not).

Indeed, checking against `address(0)` works for this

**Comment by nventuro** [comment]
So, after looking at this in some more detail, I've arrived at a tentative design, and two ways to implement said design. I believe the second approach is the way to go, but I may be missing something: input is appreciated.

The core idea is based on each `Item` being an ERC20 token, which players can access through the game and trade as usual. `Item`s would have two additional functions: a `subtract` (or `useAsIngredient`) function that removes tokens from a user, and a non-implemented `craft` function that's used to add tokens to a user. A derived `BasicItem` contract would implement `craft` by simply giving tokens to the user, and a derived `AdvancedItem` (recipe) contract (which is constructed receiving an array of `Item`s - the ingredients) implements `craft` by first removing tokens from each of the ingredient contracts. The logic is contained in the `Item`s, and the game would simply provide a way to access them (and could later on implement some other features, like charging ETH to craft and then do something with the accumulated funds, etc.)

### Using plain ERC20s
Each `Item` could be constructed with a virtually infinite supply of tokens, owned by the `Item` creator (the game). Giving tokens (from the game's supply) to a player is a non-issue, but removing them is. Each player would have to call `approve` on each `BasicItem` with the game's address, and then `approve` each `AdvancedItem` for each ingredient, resulting in a lot of transactions to start playing.

### Improving on top of ERC20s
Basically, `Item`s are similar to a mixture of Zeppelin's `MintableToken` and `BurnableToken`: the game has no token balance, tokens are simply `mint`ed by the `Item` owner (the game), and `burn`ed when used as ingredients. The issue here is once again ownership: `AdvancedItem`s will not be allowed to call `burn` on their ingredients. A solution would be to make a custom contract, based on `Mintable` and `Burnable`, where the owner (the game) can not only call `burn` on anyone's tokens, but also appoint burner addresses, which are also allowed to call `burn` (each `AdvancedItem`). 
I don't see any issue with giving the game and recipes permission to effectively destroy tokens, since they would always be the caller's tokens.

**Comment by spalladino** [comment]
Another approach that comes to mind is that `AdvancedItem`s are not the ones who actually burn their ingredients, but the game contract is. This way, a Game only needs to query an `AdvancedItem`'s recipe to see what basic ingredients are needed, burn them from the player's balance, and mint a new AdvancedItem for the player. This should simplify the permissions scheme: all tokens are dumb, and the only one with logic and permission to execute crafts is the Game itself.

**Comment by nventuro** [comment]
I thought about that, but I liked the idea of delegating the recipe behavior to the contracts: otherwise, they are just dumb tokens plus data, with no guarantee of the recipe restrictions being in place. Plus I'd avoid having to return arrays of structs from a function call, which from what I've read is a bit awkward. 

The game would still need to be allowed to burn a token in that scenario, so it comes down to whether there's one or multiple burners, and who's responsible for carrying out the recipe logic (by burning).

**Comment by spalladino** [comment]
The advantage of having just the game burning is that you only need to authorize the game contract to burn on your behalf. However, it may be possible to delegate burning to just one contract, and still have the recipes be executed by the advanced items. I'll leave you to it!

**Comment by nventuro** [comment]
Now that `Item`s are ERC20 tokens, this also closes #7.
I'm also switching to dynamic recipes, since the contract with hardcoded tokens is too expensive to be deployed at once.

**Comment by nventuro** [comment]
In the end, I went for the items (now `CraftableToken`s) providing storage for the recipe, but having the actual burning be done by the game contract. This way, the game can still `mint` new tokens with no recipe restrictions, allowing them to be acquired by other means (e.g. purchasing, random bundles, etc.). Also, because the game is always the owner of the `CraftableToken`s, it can burn them without the need of receiving authorization from the players.

I also added a new migration step with the adding of tokens and ingredients according to `rules.json`, in such a way that the tests can easily replicate it and therefore test the actual deployed contract (is there a way to call `truffle migrate` from the test suite?).

@spalladino PTAL.

File: contracts/CraftableToken.sol [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    Mint(player, amount);
+    Transfer(address(0), player, amount);
+  }
+
+  /**
+   * @dev Burns (destroys) tokens from a player's supply.
+   * @param player The player whose tokens will be burned.
+   * @param amount The amount of tokens to burn.
+   */
+  function burn(address player, uint256 amount) public onlyOwner {
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/CraftableToken.sol]: OpenZeppelin's `BurnableToken` doesn't quite work like that: token holders can burn their token, but the token contract owner cannot do anything. There's actually an [open issue](https://github.com/OpenZeppelin/zeppelin-solidity/issues/785) suggesting changing this to mimic how `MintableToken` works.

File: contracts/CraftableToken.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  /**
+   * @dev Returns the amount of CraftableTokens that must be burned to comply
+   * with a recipe step.
+   * @param recipeStep The queried recipe step.
+   */
+  function getAmountNeeded(uint256 recipeStep) public view validStep(recipeStep) returns (uint256) {
+    return recipe[recipeStep].amountNeeded;
+  }
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/CraftableToken.sol]: I originally tried to return a struct, only to find those are not allowed as return types for public functions. Must've missed the bit about returning tuples, thanks!

File: contracts/CraftableToken.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      amountNeeded: amountNeeded
+    }));
+  }
+
+  /**
+   * @dev Mints (creates) new tokens and adds it to a player's supply.
+   * @param player The player that will receive the minted tokens.
+   * @param amount The amount of tokens to mint.
+   */
+  function mint(address player, uint256 amount) public onlyOwner {
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/CraftableToken.sol]: `Mintable` would work fine in this case, but it also has added functionality that I don't need (the `stopMinting` function). I opted for removing those bits to make the contract smaller/calls cheaper.

File: contracts/Crafty.sol [hunk 4]
```diff
@@ -1,123 +1,102 @@
...
+  }
+
+  /**
+   * @dev Crafts a craftable. All of the craftable's ingredients must be
+   * present in the player's inventory in the required amounts, and they
+   * will be consumed (destroyed) by the crafting process.
+   * @param name The name of the craftable to craft.
+   */
+  function craft(string name) public {
+    CraftableToken result = getCraftable(name);
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/Crafty.sol]: I was thinking along the lines of a recipe, which has ingredients, steps, and a... result? I'm not a fan of the name either, but couldn't come up with anything better. `dish` seems like going too far with the analogy, `name` is a bit generic, `target` is plain weird. I'm open to suggestions!

File: contracts/Crafty.sol [hunk 5]
```diff
@@ -1,123 +1,102 @@
...
-    require(amountBronze() >= 3);
+    // Check all required items are present in the player's inventory.
+    for (i = 0; i < totalSteps; ++i) {
+      require(result.getIngredient(i).balanceOf(player) >= result.getAmountNeeded(i));
+    }
 
-    address player = msg.sender;
+    // After the check passed, subtract those items. We loop twice to prevent
+    // subtracting item until we know the subtract call won't fail (and cause
+    // items to be permanently lost).
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/Crafty.sol]: Even state changes in another contract? I thought that wasn't the case, thanks for the info!

File: contracts/CraftableToken.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/CraftableToken.sol]: There's also the fact that I had to reimplement `BurnableToken`, and I liked the symmetry and clarity of having them both there.

File: contracts/CraftableToken.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (spalladino) [reply contracts/CraftableToken.sol]: Good point!
